data-prepper-api/src/main/java/org/opensearch/dataprepper/model/annotations/DataPrepperPlugin.java

```diff
@@ -72,4 +72,16 @@
      * @since 1.2
      */
     Class<?> pluginConfigurationType() default PluginSetting.class;
+
+    /**
+     * Optional Packages to scan for Data Prepper DI components.
+     * Plugins provide this list if they want to use Dependency Injection in its module.
+     * Providing this value, implicitly assumes and initiates plugin specific isolated ApplictionContext.
+     * <p>
+     * The package names that spring context scans will be picked up by these marker classes.
+     *
+     * @return Array of packages to scan
+     * @since 2.2
+     */
+    Class[] packagesToScanForDI() default {};
```

Comment:
@chenqi0805 , What do you mean by the runtime dependencies?

Overall, I think this approach will work and keep the packages decoupled.

@san81 , Let's use a different name here. Perhaps `packagesToScan` or `componentsPackages`.

Reply:
My concern is if a plugin needs to scan a particular package say "org.opensearch.dataprepper.custom" but the module is not specified in the plugin build.gradle runtime dependencies, it will cause runtime exception. But it might not be a big issue depending on what packages we allow plugins to scan.

Reply:
It is certainly possible to have runtime errors here. Something we probably need in general is a nice framework for running plugins in isolation to make sure all the DI is available.

For example, factories like this could be loaded from DI:

https://github.com/opensearch-project/data-prepper/blob/f6a06a0a43b1912cf9b37a14353e48dc19fb85a0/data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/s3/S3Source.java#L79

This is a dependency that the plugin current creates by calling `new`, which is not ideal.

The typical use-case here is that a plugin is creating its own packages to include. But, this does raise an interesting point. We may wish to limit the packages that the plugin can use. For example, this could be used to incorrectly load Data Prepper core DI.